File: Python/Arrays/binary_search.py
# https://www.geeksforgeeks.org/binary-search/

import sys


class Solution:
    ##Complete this function
    def searchInSorted(self, arr, N, K):
        # Your code here
        i = 0
        j = N - 1

        while i < j:
            m = (i + j) // 2

            if arr[m] > K:
                j = m - 1
            elif arr[m] < K:
                i = m + 1
            else:
                return 1

        if arr[i] == K:
            return i
        else:
            return -1


if __name__ == "__main__":
    print(Solution().searchInSorted([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 10, 4))

File: cpp/array/binary_search.cpp
#include<iostream>
#include<vector>

using namespace std;

int solution(vector<int>& A, int X) {
    int N = A.size();
    if (N == 0) {
        return -1;
    }
    int l = 0;
    int r = N - 1;
    while (l < r) {
        // printf("l %d r %d", l, r);
        
        int m = (l + r) / 2;
        if (A[m] > X) {
            r = m - 1;
        } else {
            if(A[m] == X) return m;
            else
                l = m + 1;
        }
    }
    if (A[l] == X) {
        return l;
    }
    return -1;
}

int main() {

    vector<int> t1 = {1, 2, 5, 9, 9};

    printf("%d", solution(t1, 2));
    return 0;
}